internal/enginenetx/httpsdialer.go

```diff
@@ -0,0 +1,530 @@
+		tactic.OnTLSHandshakeError(err)
+		tcpConn.Close()
+		return nil, err
+	}
+
+	// verify the certificate chain
+	ol = logx.NewOperationLogger(logger, "TLSVerifyCertificateChain %s", tactic.VerifyHostname())
+	err = httpsDialerVerifyCertificateChain(tactic.VerifyHostname(), tlsConn, hd.rootCAs)
+	ol.Stop(err)
+
+	// handle verification error
+	if err != nil {
+		tactic.OnTLSVerifyError(err)
+		tlsConn.Close()
+		return nil, err
+	}
+
+	// make sure the tactic know it worked
+	tactic.OnSuccess()
+
+	return tlsConn, nil
+}
+
+// httpsDialerWaitReady waits for the given delay to expire or the context to be canceled. If the
+// delay is zero or negative, we immediately return nil. We also return nil when the delay expires. We
+// return the context error if the context expires.
+func httpsDialerTacticWaitReady(ctx context.Context, tactic HTTPSDialerTactic) error {
+	delay := tactic.InitialDelay()
+	if delay <= 0 {
+		return nil
+	}
+
+	timer := time.NewTimer(delay)
+	defer timer.Stop()
+
+	select {
+	case <-timer.C:
+		return nil
+
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// errNoPeerCertificate is an internal error returned when we don't have any peer certificate.
+var errNoPeerCertificate = errors.New("no peer certificate")
+
+// httpsDialerVerifyCertificateChain verifies the certificate chain with the given hostname.
+func httpsDialerVerifyCertificateChain(hostname string, conn model.TLSConn, rootCAs *x509.CertPool) error {
+	// This code comes from the example in the Go source tree that shows
+	// how to override certificate verification and which is advertised
+	// as follows:
+	//
+	//	VerifyConnection can be used to replace and customize connection
+	//	verification. This example shows a VerifyConnection implementation that
+	//	will be approximately equivalent to what crypto/tls does normally to
+	//	verify the peer's certificate.
+	//
+	// See https://github.com/golang/go/blob/go1.21.0/src/crypto/tls/example_test.go#L186
+	//
+	// As of go1.21.0, the code we're replacing has approximately the same
+	// implementation of the verification code we added below.
+	//
+	// See https://github.com/golang/go/blob/go1.21.0/src/crypto/tls/handshake_client.go#L962.
+
+	state := conn.ConnectionState()
+	opts := x509.VerifyOptions{
+		DNSName:       hostname, // note: here we're using the real hostname
+		Intermediates: x509.NewCertPool(),
+		Roots:         rootCAs,
+	}
+
+	// The following check is rather paranoid and it's not part of the Go codebase
+	// from which we copied it, but I think it's important to be defensive.
+	//
+	// Because of that, I don't want to just drop an assertion here.
+	if len(state.PeerCertificates) < 1 {
+		return errNoPeerCertificate
+	}
+
+	for _, cert := range state.PeerCertificates[1:] {
+		opts.Intermediates.AddCert(cert)
+	}
+
+	if _, err := state.PeerCertificates[0].Verify(opts); err != nil {
+		return netxlite.NewErrWrapper(netxlite.ClassifyTLSHandshakeError, netxlite.TopLevelOperation, err)
+	}
+	return nil
```

Comment:
For comparison, this is what Go code does:

```Go
	if !c.config.InsecureSkipVerify {
		opts := x509.VerifyOptions{
			Roots:         c.config.RootCAs,
			CurrentTime:   c.config.time(),
			DNSName:       c.config.ServerName,
			Intermediates: x509.NewCertPool(),
		}

		for _, cert := range certs[1:] {
			opts.Intermediates.AddCert(cert)
		}
		var err error
		c.verifiedChains, err = certs[0].Verify(opts)
		if err != nil {
			c.sendAlert(alertBadCertificate)
			return &CertificateVerificationError{UnverifiedCertificates: certs, Err: err}
		}
	}
```

See https://github.com/golang/go/blob/go1.21.0/src/crypto/tls/handshake_client.go#L962